File: approaches/lifetime-objective/leaf-evolution/decide.cpp
// One-shot native decision for the benchmark playground and the competition.
//
//   decide --board <49 digits> --next <1-7> --rise <1-5>
//          [--weights file] [--depth 4] [--chance-samples 7] [--cache 60000]
//
// prints "bestmove <column>" (0-6) or "bestmove none" on a terminal board, and
// exits 0.  The board is the engine's serializeBoard encoding, row-major from
// the top: 0 empty, 1-7 numbered, 8 solid gray, 9 cracked gray -- the same
// string the D7P protocol carries (docs/d7p-protocol.md).
//
// The policy reads exactly the public state: visible board, visible next disc,
// moves until the next rise.  There is no seed, score, level or move number on
// the command line, so there is nothing else it could read.  Decisions are
// deterministic for a given board, weights and configuration, which the
// benchmark harness requires.
//
// With no --weights this is the frozen fast fair-D4 leaf; with a weights file
// it is the same search with that leaf.  The search is the gated
// WeightedFastSearch (see build.sh), so a column chosen here is the column the
// research evaluator would choose on the same public state.

#include "weighted-search.hpp"

#include <cstring>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

using namespace drop7;

std::uint64_t worstCaseWork(int maximumDepth, int strata) {
  const auto branches =
      static_cast<std::uint64_t>(kBoardSize) * static_cast<std::uint64_t>(strata);
  std::uint64_t total = 0;
  for (int depth = 1; depth <= maximumDepth; ++depth) {
    std::uint64_t power = 1;
    for (int level = 1; level <= depth; ++level) {
      power *= branches;
      total += power;
    }
    total += power;
  }
  return total;
}

}  // namespace

int main(int argc, char** argv) {
  try {
    std::string board;
    std::string weightsPath;
    int next = 0;
    int rise = 0;
    fastw::FastSearchParameters parameters;
    parameters.depth = 4;
    parameters.chance_samples = 7;
    parameters.maximum_cache_entries = 60'000;
    for (int i = 1; i < argc; ++i) {
      const std::string key = argv[i];
      if (i + 1 >= argc) throw std::invalid_argument("missing value for " + key);
      const std::string value = argv[++i];
      if (key == "--board") board = value;
      else if (key == "--next") next = std::stoi(value);
      else if (key == "--rise") rise = std::stoi(value);
      else if (key == "--weights") weightsPath = value;
      else if (key == "--depth") parameters.depth = std::stoi(value);
      else if (key == "--chance-samples") parameters.chance_samples = std::stoi(value);
      else if (key == "--cache") parameters.maximum_cache_entries = std::stoull(value);
      else throw std::invalid_argument("unknown option " + key);
    }
    if (board.size() != static_cast<std::size_t>(kCellCount)) {
      throw std::invalid_argument("--board must be 49 characters");
    }
    if (next < 1 || next > kBoardSize) throw std::invalid_argument("--next must be 1-7");
    if (rise < 1 || rise > kMovesPerLevel) throw std::invalid_argument("--rise must be 1-5");
    parameters.maximum_work = worstCaseWork(parameters.depth, parameters.chance_samples) + 1;

    State state;
    for (int cell = 0; cell < kCellCount; ++cell) {
      const char c = board[static_cast<std::size_t>(cell)];
      if (c < '0' || c > '9') throw std::invalid_argument("--board cells must be digits 0-9");
      state.board[static_cast<std::size_t>(cell)] = static_cast<std::uint8_t>(c - '0');
    }
    state.next_disc = static_cast<std::uint8_t>(next);
    state.moves_remaining = rise;
    state.game_over = false;

    fastw::LeafWeights weights;
    if (!weightsPath.empty()) weights = fastw::readWeightsFile(weightsPath);

    fastw::WeightedFastSearch search{parameters, weights};
    fastw::FastSearchMetrics metrics;
    const int action = search.chooseAction(state, metrics);
    if (action < 0 || !isLegal(state.board, action)) {
      std::cout << "bestmove none\n";
    } else {
      std::cout << "bestmove " << action << "\n";
    }
    std::cout << "info depth " << metrics.completed_depth << " work " << metrics.work
              << " nodes " << metrics.nodes << " frozen " << (weights.isFrozen() ? 1 : 0) << "\n";
    return 0;
  } catch (const std::exception& error) {
    std::cerr << "decide failed: " << error.what() << '\n';
    return 2;
  }
}

File: approaches/fair-expectimax/rust-engine/src/bin/decide.rs
// One-shot decision binary for the benchmark playground and the competition,
// the Rust counterpart of approaches/lifetime-objective/leaf-evolution's
// decide.cpp:
//
//   decide --board <49 digits> --next <1-7> --rise <1-5>
//          [--depth 7] [--chance-samples 7]
//          [--cache TOTAL_ENTRIES] [--threads T]
//          [--scheduler frontier|root] [--split-plies auto|N]
//          [--max-frontier-tasks N] [--max-host-bytes N]
//
// prints "bestmove <column>" (0-6) or "bestmove none" on a terminal board, and
// exits 0.  The board is the engine's serializeBoard encoding, row-major from
// the top: 0 empty, 1-7 numbered, 8 solid gray, 9 cracked gray -- the same
// string the D7P protocol carries (docs/d7p-protocol.md).
//
// The policy reads exactly the public state: visible board, visible next disc,
// moves until the next rise.  There is no seed, score, level or move number on
// the command line, so there is nothing else it could read.  Decisions are
// deterministic for a given board and configuration, which the benchmark
// harness requires.
//
// CENTRAL-FRONTIER PARALLELISM IS VALUE-IDENTICAL, NOT AN APPROXIMATION.  A
// coordinator expands a deterministic prefix of the expectimax tree, workers
// claim the remaining public-state continuations from a central registry, and
// the coordinator reduces results in the original column/sample order.  This
// exposes thousands of tasks instead of at most seven root columns.  Private
// worker tables can change cache hits and logical work, but cached values are
// bit-identical to recomputation, so threads change no root value or action.
// `--scheduler root` retains the earlier coarse-grained implementation as a
// measured fallback. `--cache` remains one aggregate entry budget, partitioned
// across private worker tables so increasing the thread count cannot multiply
// the caller's memory request.

use drop7_rs::board::{Board, BOARD_SIZE};
use drop7_rs::engine::State;
use drop7_rs::parallel::{
    choose_action_frontier_parallel, choose_action_root_parallel, ParallelConfig,
    DEFAULT_MAX_FRONTIER_TASKS, DEFAULT_MAX_HOST_BYTES,
};
use drop7_rs::search::{work_bound_for, SearchParams};

use std::env;

fn main() {
    let mut board_text = String::new();
    let mut next = 0u8;
    let mut rise = 0i32;
    let mut depth = 7i32;
    let mut strata = 7i32;
    let mut cache = 1_048_576usize;
    let mut threads = 0usize; // 0 = all CPUs visible to this process
    let mut scheduler = String::from("frontier");
    let mut split_plies = None;
    let mut max_frontier_tasks = DEFAULT_MAX_FRONTIER_TASKS;
    let mut max_host_bytes = DEFAULT_MAX_HOST_BYTES;
    let args: Vec<String> = env::args().collect();
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--help" || args[i] == "-h" {
            eprintln!("usage: decide --board BOARD --next 1-7 --rise 1-5 [--depth N] [--chance-samples N] [--cache N] [--threads N] [--scheduler frontier|root] [--split-plies auto|N] [--max-frontier-tasks N] [--max-host-bytes N]");
            return;
        }
        if i + 1 >= args.len() {
            eprintln!("decide failed: {} needs a value", args[i]);
            std::process::exit(2);
        }
        match args[i].as_str() {
            "--board" => board_text = args[i + 1].clone(),
            "--next" => next = args[i + 1].parse().expect("--next must be 1-7"),
            "--rise" => rise = args[i + 1].parse().expect("--rise must be 1-5"),
            "--depth" => depth = args[i + 1].parse().expect("--depth"),
            "--chance-samples" => strata = args[i + 1].parse().expect("--chance-samples"),
            "--cache" => cache = args[i + 1].parse().expect("--cache"),
            "--threads" => threads = args[i + 1].parse().expect("--threads"),
            "--scheduler" => scheduler = args[i + 1].clone(),
            "--split-plies" => {
                split_plies = if args[i + 1] == "auto" {
                    None
                } else {
                    Some(
                        args[i + 1]
                            .parse()
                            .expect("--split-plies must be auto or an integer"),
                    )
                }
            }
            "--max-frontier-tasks" => {
                max_frontier_tasks = args[i + 1].parse().expect("--max-frontier-tasks")
            }
            "--max-host-bytes" => max_host_bytes = args[i + 1].parse().expect("--max-host-bytes"),
            other => {
                eprintln!("decide failed: unknown option {other}");
                std::process::exit(2);
            }
        }
        i += 2;
    }
    if depth < 1 || strata < 1 {
        eprintln!("decide failed: depth and chance samples must be at least 1");
        std::process::exit(2);
    }
    let Some(board) = Board::from_serialized(&board_text) else {
        eprintln!("decide failed: --board must be 49 characters of digits 0-9");
        std::process::exit(2);
    };
    if !(1..=BOARD_SIZE as u8).contains(&next) {
        eprintln!("decide failed: --next must be 1-7");
        std::process::exit(2);
    }
    if cache == 0 {
        eprintln!("decide failed: --cache must be at least 1");
        std::process::exit(2);
    }

    // The public state only: no score, level or move number exists here, and
    // the search and leaf provably read none of them.
    let state = State {
        board,
        next_disc: next,
        score: 0,
        level: 1,
        moves_remaining: rise,
        moves_played: 0,
        game_over: false,
    };

    let params = SearchParams {
        depth,
        chance_samples: strata,
        terminal_utility: -1_000_000.0,
        maximum_work: work_bound_for(depth, strata) + 1,
        policy_seed: 0xd707_5eed,
    };

    let requested_threads = if threads == 0 {
        std::thread::available_parallelism().map_or(1, |n| n.get())
    } else {
        threads
    };
    // DepthTable rounds to powers of two. Partition the aggregate cache budget
    // before constructing ParallelConfig so all private worker tables together
    // remain at or below the caller's requested entry count.
    let worker_count = requested_threads.min(cache);
    let worker_cache = cache_entries_per_worker(cache, worker_count);
    let config = ParallelConfig {
        threads: worker_count,
        table_capacity_per_worker: worker_cache,
        split_plies,
        max_frontier_tasks,
        max_host_bytes,
        ..ParallelConfig::default()
    };
    let decision = match scheduler.as_str() {
        "frontier" => choose_action_frontier_parallel(&state, params, config),
        "root" => choose_action_root_parallel(&state, params, config),
        _ => {
            eprintln!("decide failed: --scheduler must be frontier or root");
            std::process::exit(2);
        }
    };
    let decision = match decision {
        Ok(decision) => decision,
        Err(error) if error.contains("no legal columns") => {
            println!("bestmove none");
            return;
        }
        Err(error) => {
            eprintln!("decide failed: {error}");
            std::process::exit(2);
        }
    };
    println!("bestmove {}", decision.action);
    let metrics = decision.metrics;
    let allocated_cache_entries = worker_cache * metrics.worker_threads;
    println!(
        "info depth {} scheduler {} threads {} tasks {} split {} work {} nodes {} busy {:.4} wall {:.6} cache-entries {} table-bytes {} frozen 1",
        depth,
        scheduler,
        metrics.worker_threads,
        metrics.frontier_tasks,
        metrics.split_plies,
        metrics.work,
        metrics.nodes,
        metrics.worker_busy_fraction,
        metrics.wall_seconds,
        allocated_cache_entries,
        metrics.projected_table_bytes,
    );
}

/// Largest power-of-two per-worker table that keeps the aggregate allocation
/// within `total_entries`. The caller guarantees both values are positive and
/// caps `worker_count` at `total_entries`.
fn cache_entries_per_worker(total_entries: usize, worker_count: usize) -> usize {
    debug_assert!(total_entries > 0);
    debug_assert!(worker_count > 0);
    debug_assert!(worker_count <= total_entries);
    let share = total_entries / worker_count;
    1usize << (usize::BITS - 1 - share.leading_zeros())
}

#[cfg(test)]
mod tests {
    use super::cache_entries_per_worker;

    #[test]
    fn cache_budget_is_shared_across_frontier_workers() {
        let total = 16_777_216usize;
        let workers = 192usize;
        let per_worker = cache_entries_per_worker(total, workers);

        assert_eq!(per_worker, 65_536);
        assert!(per_worker.is_power_of_two());
        assert!(per_worker * workers <= total);
    }

    #[test]
    fn cache_partition_never_exceeds_the_total_budget() {
        for total in 1..=257usize {
            for workers in 1..=total.min(257) {
                let per_worker = cache_entries_per_worker(total, workers);
                assert!(per_worker.is_power_of_two());
                assert!(per_worker * workers <= total);
            }
        }
    }
}
